### PR TITLE
Fixes #12233: printing environment corrupted with eta-expansion of "match" branches

### LINKED ISSUE #12233: "Show Proof" raises "Not_found" anomaly
Adding "Show Proof" after the line `+ intuition.` in theorem `unrestricted_minimization_entails_excluded_middle` in `ClassicalFacts.v` generates a `Not_found` anomaly.

I discovered this using a modified coqc that automatically applies a `Show Proof` after every tactic, then verified the bug in the official release.  You get the same error on the following lines up to the Qed, but only in this proof in this library file.  Nowhere else.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

The main fix is in `Detyping.decomp_branch` which needs a declaration to push to the environment even in the eta-expansion cases.

In the process, we preserve binder annotations in detyping (maybe not so useful, but morally better, for instance for funind).

Note: we left an horrible hack in the case of LetIn expansion. This is supposed to disappear with #9710.

Fixes / closes #12233

- [X] Added / updated test-suite
- [x] Entry added in the changelog

### CODE REVIEW

File: kernel/context.ml
Comment (ppedrot) [inline kernel/context.ml:182]: I'd rather not have this function added. The reason we have the dual one is due to historical reasons, in order to easily port code using the previous API. Rather, the only user introduced in this PR should match on the argument.

Comment (herbelin) [reply kernel/context.ml:182]: I agree. So, just to be sure, do you mean directly passing a decl to add_name, rather than deconstructing/reconstructing?

PS: There is already an of_tuple for named decl!

Comment (ppedrot) [reply kernel/context.ml]: @herbelin yes, the caller code seems easily amenable to using constructors directly.

File: Makefile.common
Comment (ppedrot) [inline Makefile.common:171]: What's the matter with this change?

**Comment by herbelin** [comment]
The fix met a "hack" of `ssrmatching.ml` which consists in sending terms with an unbound variable called `_` to `pr_constr`. The purpose of the ssreflect hack is to print evars `_` rather than `?x`.

How much is this "feature" important? If printing `_` instead of `?x` for some users is important why not simply adding a flag modifying `constrextern` using meaningful constructions  rather than building `constr` with invalid (and undeclared) variable names?

In the meantime,  I added a hack on top of the hack, but I believe a more honorable implementation would really be worth (if ever the feature itself is really worth).

File: Makefile.common
Comment (herbelin) [reply Makefile.common]: This is the code I use to debug plugins. I commited it by mistake (now removed).

Note in passing that this shows that something is wrong with the makefile which fails on the Ltac2 plugin when dynlink is not available.

**Comment by ppedrot** [comment]
I'd like a look from @coq/ssreflect-maintainers before merging. This is fairly urgent if this is supposed to make it into 8.11.2 since I need to tag before the 8.12 branch supposed to occur on Friday. That said, this is a relatively minor bug, so I guess it's also fine if not backported.

**Comment by ejgallego** [comment]
> I'd like a look from @coq/ssreflect-maintainers before merging.

Ping @coq/ssreflect-maintainers

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
Changes to SSR seem fine, but I've no idea what the rest of the patch does

**Pull request merged**
